File: client/common/protocol.go
package common

import (
	"encoding/binary"
	"fmt"
)

// Protocol Entity
type Protocol struct {
	id int
}

/*
 * Protocol definition:

 * TLV (Type, Length, Value) format
	* Type: 1 byte
	* Length: 1 byte (or 2 bytes for Batch type)
	* Value: n bytes

 * Protocol types:
	* 0x01: Batch (n bytes, max 65535, composed of the following fields)
		* 0x01: Amount of bets (uint32, 4 bytes)
		* 0x02: Bet (n bytes as defined below, could be multiple)
		* 0x03: Agency ID (uint32, 4 bytes)
		* 0x04: Last batch flag (1 byte)
	* 0x02: Bet (n bytes, composed of the following fields)
		* 0x01: Name (string, n bytes, max 255)
		* 0x02: Surname (string, n bytes, max 255)
		* 0x03: DNI (uint32, 4 bytes)
		* 0x04: Birthdate (string, n bytes, max 255)
		* 0x05: Number (uint32, 4 bytes)
	* 0x03: Response (1 byte)
	* 0x04: Agency ID (4 bytes)
	* 0x05: Winners (n bytes, max 65535, composed of the following fields)
		* 0x01: Winner (n bytes, could be 0 or multiple)
			* 0x01: DNI (uint32, 4 bytes)
			* 0x02: Number (uint32, 4 bytes)
*/

// CreateBetBatchMessage: Creates a batch message with the given list of bets and EOF flag using TLV (type, length, value) format
func (p *Protocol) CreateBetBatchMessage(bets []Bet, eof bool) []byte {
	// Create the message as a byte slice
	msg := make([]byte, 0)
	serializedBets := make([]byte, 0)
	// Start the length at 9 (4 bytes for the agency ID, 1 byte for the last batch flag
	// and 4 bytes for the amount of bets)
	length := 9
	for _, bet := range bets {
		// Calculate the length of the bet message
		length += len(bet.Name) + len(bet.Surname) + len(bet.Birthdate) + 8
		// Create the bet message
		betMsg := p.CreateBetMessage(bet.Name, bet.Surname, bet.DNI, bet.Birthdate, bet.Number)
		// Append the bet message
		serializedBets = append(serializedBets, betMsg...)
	}

	// Append the TLV fields

	// Type (Batch)
	msg = append(msg, 0x01)
	lengthBytes := make([]byte, 2)
	binary.BigEndian.PutUint16(lengthBytes, uint16(length))
	msg = append(msg, lengthBytes...)

	// Amount of bets (4 bytes), we put it first so it's the first field to be read
	amount := make([]byte, 4)
	binary.BigEndian.PutUint32(amount, uint32(len(bets)))
	msg = append(msg, 0x01)
	msg = append(msg, byte(len(amount)))
	msg = append(msg, amount...)

	// Bets (append all the serialized bets)
	msg = append(msg, serializedBets...)

	// Agency ID (4 bytes)
	agencyID := make([]byte, 4)
	binary.BigEndian.PutUint32(agencyID, uint32(p.id))
	msg = append(msg, 0x03)
	msg = append(msg, byte(len(agencyID)))
	msg = append(msg, agencyID...)

	// Last batch flag (1 byte)
	flag := byte(0)
	if eof {
		flag = 1
	}
	msg = append(msg, 0x04)
	msg = append(msg, 1)
	msg = append(msg, flag)

	return msg
}

// CreateBetMessage: Creates a bet message with the given parameters using TLV (type, length, value) format
func (p *Protocol) CreateBetMessage(
	name string,
	surname string,
	dni int,
	birthdate string,
	number int) []byte {
	// Create the message as a byte slice
	msg := make([]byte, 0)
	length := len(name) + len(surname) + len(birthdate) + 8

	// Append the TLV fields

	// Type (Bet)
	msg = append(msg, 0x02)
	msg = append(msg, byte(length))

	// Name
	msg = append(msg, 0x01)
	msg = append(msg, byte(len(name)))
	msg = append(msg, []byte(name)...)

	// Surname
	msg = append(msg, 0x02)
	msg = append(msg, byte(len(surname)))
	msg = append(msg, []byte(surname)...)

	// DNI
	msg = append(msg, 0x03)
	dniBytes := make([]byte, 4)
	binary.BigEndian.PutUint32(dniBytes, uint32(dni))
	msg = append(msg, byte(len(dniBytes)))
	msg = append(msg, dniBytes...)

	// Birthdate
	msg = append(msg, 0x04)
	msg = append(msg, byte(len(birthdate)))
	msg = append(msg, []byte(birthdate)...)

	// Number
	msg = append(msg, 0x05)
	numberBytes := make([]byte, 4)
	binary.BigEndian.PutUint32(numberBytes, uint32(number))
	msg = append(msg, byte(len(numberBytes)))
	msg = append(msg, numberBytes...)

	return msg
}

func (p *Protocol) CreateAgencyIDMessage() []byte {
	// Create the message as a byte slice
	msg := make([]byte, 0)

	// Append the TLV fields

	// Type (Agency ID)
	msg = append(msg, 0x04)
	agencyID := make([]byte, 4)
	binary.BigEndian.PutUint32(agencyID, uint32(p.id))
	msg = append(msg, byte(len(agencyID)))
	msg = append(msg, agencyID...)

	return msg
}

// ParseResponse: Parses the response message and returns the result
func (p *Protocol) ParseResponse(msg []byte) error {
	// The response will use the TLV format but its value will be a single byte long
	// This is trivial now and could be done without the TLV format, but it's done so it
	// can easily be extended in the future to handle more information.
	if len(msg) != 3 {
		return fmt.Errorf("invalid response message")
	}

	// Check the type (should be response)
	if msg[0] != 0x03 {
		return fmt.Errorf("invalid response type")
	}

	// Check the length (should be 1)
	if msg[1] != 1 {
		return fmt.Errorf("invalid response length")
	}

	// Check the value (0 for failure, 1 for success)
	if msg[2] != 1 {
		return fmt.Errorf("server returned an error")
	}

	return nil
}

// ParseWinners: Parses the winners message and returns the list of winners
func (p *Protocol) ParseWinners(msg []byte) ([]Winner, error) {
	// The winners message will use the TLV format, with the value being a list of winners
	// Each winner will have a DNI and a number
	if len(msg) < 3 {
		return nil, fmt.Errorf("invalid winners message")
	}

	// Check the type (should be winners)
	if msg[0] != 0x05 {
		return nil, fmt.Errorf("invalid winners type")
	}

	// Check the length (next 2 bytes)
	length := int(binary.BigEndian.Uint16(msg[1:3]))

	if length == 0 {
		// No winners
		return []Winner{}, nil
	}

	// Parse the winners
	winners := make([]Winner, 0)

	// Start at the first winner
	i := 3
	// Loop until the end of the message
	for i < len(msg) {
		// Check if there's a winner field
		if msg[i] != 0x01 {
			return nil, fmt.Errorf("invalid winner type")
		}
		// Check if length is correct
		if msg[i+1] != 8 {
			return nil, fmt.Errorf("invalid winner length")
		}
		// Check if there's a DNI field
		if msg[i+2] != 0x01 {
			return nil, fmt.Errorf("invalid winner DNI type")
		}
		// Check DNI length
		dniLength := int(msg[i+3])
		if dniLength != 4 {
			return nil, fmt.Errorf("invalid winner DNI length")
		}
		dni := binary.BigEndian.Uint32(msg[i+4 : i+8])
		// Check if there's a number field
		if msg[i+8] != 0x02 {
			return nil, fmt.Errorf("invalid winner number type")
		}
		// Check number length
		numberLength := int(msg[i+9])
		if numberLength != 4 {
			return nil, fmt.Errorf("invalid winner number length")
		}
		number := binary.BigEndian.Uint32(msg[i+10 : i+14])
		// Append the winner
		winners = append(winners, Winner{DNI: dni, Number: number})
		// Move to the next winner
		i += 14
	}

	return winners, nil
}

File: server/common/protocol.py
import logging
from common.utils import Bet

"""
Protocol class

This class is responsible for parsing the messages received from the client
and creating the responses.

The messages are expected to follow a TLV (Type, Length, Value) format where
each field is indicated by a type byte, a length byte and the value itself.

    * TLV (Type, Length, Value) format
        * Type: 1 byte
        * Length: 1 byte (or 2 bytes for Batch type)
        * Value: n bytes

    * Protocol types:
        * 0x01: Batch (n bytes, max 65535, composed of the following fields)
            * 0x01: Amount of bets (uint32, 4 bytes)
            * 0x02: Bet (n bytes as defined above, could be multiple)
            * 0x03: Agency ID (uint32, 4 bytes)
            * 0x04: Last batch flag (1 byte)
        * 0x02: Bet (n bytes, composed of the following fields)
            * 0x01: Name (string, n bytes, max 255)
            * 0x02: Surname (string, n bytes, max 255)
            * 0x03: DNI (uint32, 4 bytes)
            * 0x04: Birthdate (string, n bytes, max 255)
            * 0x05: Number (uint32, 4 bytes)
        * 0x03: Response (1 byte)
"""
class Protocol:
    def __init__(self):
        pass

    def parseBatchMessage(self, message):
        """
        Parse the batch message

        The batch is expected to follow the TLV format indicated in the
        class description. There could be multiple bets, the message is 
        expected to contain at least one. For simplicity, it is expected
        that the fields are in the correct order:
            
            * Amount of bets
            * Bet 1
            * Bet 2
            * ...
            * Bet n
            * Agency ID
            * Last batch flag

        """

        logging.debug(f"action: apuesta_recibida | result: in_progress | message: {message}")

        length = 0
        # Check that the message is of type Batch
        if message[0] != 0x01:
            raise ValueError("invalid message type, expected Batch")
        # Get message length (2 bytes)
        total_length = int.from_bytes(message[1:3], byteorder='big')

        # Check that the amount of bets is next
        if message[3] != 0x01:
            raise ValueError("invalid message format, expected amount of bets first")
        # Check that the amount of bets is 4 bytes
        if message[4] != 4:
            raise ValueError("invalid message format, expected 4 bytes for amount of bets")
        # Get the amount of bets
        amount_of_bets = int.from_bytes(message[5:9], byteorder='big')
        length += 4

        # Check that the bets are next
        if message[9] != 0x02:
            logging.error(f"action: apuesta_recibida | result: fail | cantidad: {amount_of_bets}")
            raise ValueError("invalid message format, expected bets")

        bets = []
        current_byte = 9
        # Parse all the bets
        while message[current_byte] == 0x02:
            current_byte += 1
            bet_length = message[current_byte]
            current_byte += 1
            length += bet_length
            # Parse the bet and get the next byte to parse
            bet, current_byte = self._parseBet(message, current_byte, bet_length, amount_of_bets)
            bets.append(bet)
        
        # Check that the agency ID is next
        if message[current_byte] != 0x03:
            logging.error(f"action: apuesta_recibida | result: fail | cantidad: {amount_of_bets}")
            raise ValueError("invalid message format, expected agency ID")
        # Check that the agency ID is 4 bytes
        if message[current_byte + 1] != 4:
            logging.error(f"action: apuesta_recibida | result: fail | cantidad: {amount_of_bets}")
            raise ValueError("invalid message format, expected 4 bytes for agency ID")
        # Get the agency ID
        agency_id = int.from_bytes(message[current_byte + 2:current_byte + 6], byteorder='big')
        length += 4        

        # Add the agency ID to the bets
        for bet in bets:
            bet.agency = agency_id

        # Check that the last batch flag is next
        if message[current_byte + 6] != 0x04:
            logging.error(f"action: apuesta_recibida | result: fail | cantidad: {amount_of_bets}")
            raise ValueError("invalid message format, expected last batch flag")
        # Check that the last batch flag is 1 byte
        if message[current_byte + 7] != 1:
            logging.error(f"action: apuesta_recibida | result: fail | cantidad: {amount_of_bets}")
            raise ValueError("invalid message format, expected 1 byte for last batch flag")
        # Get the last batch flag
        last_batch = message[current_byte + 8]
        length += 1

        # Check that the total length of the message is correct
        if length != total_length:
            logging.error(f"action: apuesta_recibida | result: fail | cantidad: {amount_of_bets}")
            raise ValueError("invalid message format, incorrect length")
        
        return bets, last_batch     


    def _parseBet(self, message, current_byte, total_length, amount_of_bets):
        """
        Parse the bet message

        The bet is expected to follow the TLV format indicated in the
        class description.
        """

        fields = {}
        start = current_byte
        length = 0

        #logging.debug(f"action: apuesta_recibida | result: in_progress | current_byte: {current_byte} | total_length: {total_length}")

        # Parse all the fields
        while start < current_byte + total_length + 10:
            field_type, field_value, field_length = self._get_next_bet_field(message, start)
            fields[field_type] = field_value# Add the length of the field's value
            length += field_length
            # Move to the next field
            start += 2 + field_length

        #logging.debug(f"action: apuesta_recibida | result: in_progress | fields: {fields}")

        # Check that all the fields are present
        if len(fields) != 5 or length != total_length:
            raise ValueError("invalid message format")
        
        try:
            # Return the bet with placeholder agency 0 (will be replaced later by the actual agency
            # ID sent in the batch message)
            return (Bet(0,fields[0x01], fields[0x02], fields[0x03], fields[0x04], fields[0x05]), start)
        except KeyError:
            logging.error(f"action: apuesta_recibida | result: fail | cantidad: {amount_of_bets}")
            raise ValueError("missing fields in Bet message")

    def _get_next_bet_field(self, message, start):
        """
        Get the next field from the bet

        The field is expected to follow the TLV format indicated in the
        class description.
        """
        # Get field type
        field_type = message[start]
        # Get field length
        field_length = message[start + 1]
        # Get field value
        field_value = message[start + 2:start + 2 + field_length]
    
        if field_type == 0x03 or field_type == 0x05:
            field_value = int.from_bytes(field_value, byteorder='big')
        else:
            field_value = field_value.decode('utf-8')
        return field_type, field_value, field_length
    
    def createResponse(self, success):
        """
        Create a response message to be sent to the client

        The response message uses TLV format with a single field of type
        0x02 indicating the result of the operation.
        
        This is trivial now and could be done without the TLV format,
        but it's done so it can easily be extended in the future to
        handle more information.
        
	    TODO: Handle more complex responses in future iterations
        """
        return bytes([0x03, 0x01, 0x01 if success else 0x00])
